payjoin/src/receive/mod.rs

```
@@ -437,45 +530,74 @@ impl ProvisionalProposal {
+        // Insert contributions at random indices for privacy
         let mut rng = rand::thread_rng();
-        let index = rng.gen_range(0..=self.payjoin_psbt.unsigned_tx.input.len());
-        self.payjoin_psbt
-            .inputs
-            .insert(index, bitcoin::psbt::Input { witness_utxo: Some(txo), ..Default::default() });
-        self.payjoin_psbt.unsigned_tx.input.insert(
-            index,
-            bitcoin::TxIn {
-                previous_output: outpoint,
-                sequence: original_sequence,
-                ..Default::default()
-            },
-        );
-    }
+        let mut inputs_amount = Amount::ZERO;
+        for (outpoint, txo) in inputs.into_iter() {
+            inputs_amount += txo.value;
+            let index = rng.gen_range(0..=self.payjoin_psbt.unsigned_tx.input.len());
+            payjoin_psbt.inputs.insert(
+                index,
+                bitcoin::psbt::Input { witness_utxo: Some(txo), ..Default::default() },
+            );
+            payjoin_psbt.unsigned_tx.input.insert(
+                index,
+                bitcoin::TxIn {
+                    previous_output: outpoint,
+                    sequence: original_sequence,
+                    ..Default::default()
+                },
+            );
+        }
 
-    pub fn is_output_substitution_disabled(&self) -> bool {
-        self.params.disable_output_substitution
-    }
+        // Add the receiver change amount to the receiver change output, if applicable
+        let min_input_amount = self.min_input_amount();
+        if inputs_amount >= min_input_amount {
+            let change_amount = inputs_amount - min_input_amount;
+            let vout_to_augment =
+                self.owned_vouts.choose(&mut rand::thread_rng()).expect("owned_vouts is empty");
+            payjoin_psbt.unsigned_tx.output[*vout_to_augment].value += change_amount;
+        } else {
+            todo!("Input amount is not enough to cover additional output value");
```

This should probably return an Error